AWS/GCE: Spread PetSet volume creation across zones, create GCE volumes in non-master zones

justinsb opened the pull request:
Long term we plan on integrating this into the scheduler, but in the
short term we use the volume name to place it onto a zone.

We hash the volume name so we don't bias to the first few zones.

If the volume name "looks like" a PetSet volume name (ending with
-<number>) then we use the number as an offset.  In that case we hash
the base name.

justinsb committed:
GCE/AWS: Spread PetSet volume creation across zones

Long term we plan on integrating this into the scheduler, but in the
short term we use the volume name to place it onto a zone.

We hash the volume name so we don't bias to the first few zones.

If the volume name "looks like" a PetSet volume name (ending with
-<number>) then we use the number as an offset.  In that case we hash
the base name.

Fixes #27256

justinsb commented:
Working on AWS...

```
I0617 19:22:47.328898       6 aws.go:954] Found instances in zones map[us-west-2a:{} us-west-2b:{} us-west-2c:{}]
I0617 19:22:47.328928       6 util.go:223] Detected PetSet-style volume name "datadir-rd-2"; index=2
I0617 19:22:47.328944       6 util.go:236] Creating volume with name "datadir-rd-2"; chose zone="us-west-2a" from zones=["us-west-2a" "us-west-2b" "us-west-2c"]
I0617 19:22:47.330082       6 aws.go:954] Found instances in zones map[us-west-2a:{} us-west-2b:{} us-west-2c:{}]
I0617 19:22:47.330101       6 util.go:223] Detected PetSet-style volume name "datadir-rd-0"; index=0
I0617 19:22:47.330117       6 util.go:236] Creating volume with name "datadir-rd-0"; chose zone="us-west-2b" from zones=["us-west-2a" "us-west-2b" "us-west-2c"]
I0617 19:22:47.354228       6 aws.go:954] Found instances in zones map[us-west-2a:{} us-west-2b:{} us-west-2c:{}]
I0617 19:22:47.354246       6 util.go:223] Detected PetSet-style volume name "datadir-rd-1"; index=1
I0617 19:22:47.354254       6 util.go:236] Creating volume with name "datadir-rd-1"; chose zone="us-west-2c" from zones=["us-west-2a" "us-west-2b" "us-west-2c"]
```

```
NAME                                       CAPACITY   ACCESSMODES   STATUS    CLAIM                  REASON    AGE       LABELS
pvc-df9c59ff-34c0-11e6-a694-02263499474d   20Gi       RWO           Bound     default/datadir-rd-0             2m        failure-domain.beta.kubernetes.io/region=us-west-2,failure-domain.beta.kubernetes.io/zone=us-west-2b
pvc-dfa1fd39-34c0-11e6-a694-02263499474d   20Gi       RWO           Bound     default/datadir-rd-1             2m        failure-domain.beta.kubernetes.io/region=us-west-2,failure-domain.beta.kubernetes.io/zone=us-west-2c
pvc-dfa32aea-34c0-11e6-a694-02263499474d   20Gi       RWO           Bound     default/datadir-rd-2             2m        failure-domain.beta.kubernetes.io/region=us-west-2,failure-domain.beta.kubernetes.io/zone=us-west-2a
```

justinsb committed:
GCE Multizone: Allow volumes to be created in non-master zone

We had a long-lasting bug which prevented creation of volumes in
non-master zones, because the cloudprovider in the volume label
admission controller is not initialized with the multizone setting
(issue #27656).

This implements a simple workaround: if the volume is created with the
failure-domain zone label, we look for the volume in that zone.  This is
more efficient, avoids introducing a new semantic, and allows users (and
the dynamic provisioner) to create volumes in non-master zones.

Fixes #27657

justinsb commented:
And it works on GCE also now... multiple bugs being fixed tonight!

```
I0618 04:07:15.910998       5 util.go:223] Detected PetSet-style volume name "datadir-rd-2"; index=2
I0618 04:07:15.911034       5 util.go:236] Creating volume for PVC "datadir-rd-2"; chose zone="us-central1-a" from zones=["us-central1-a" "us-central1-b" "us-central1-f"]
I0618 04:07:15.911949       5 util.go:223] Detected PetSet-style volume name "datadir-rd-0"; index=0
I0618 04:07:15.911982       5 util.go:236] Creating volume for PVC "datadir-rd-0"; chose zone="us-central1-b" from zones=["us-central1-a" "us-central1-b" "us-central1-f"]
I0618 04:07:15.913963       5 util.go:223] Detected PetSet-style volume name "datadir-rd-1"; index=1
I0618 04:07:15.913998       5 util.go:236] Creating volume for PVC "datadir-rd-1"; chose zone="us-central1-f" from zones=["us-central1-a" "us-central1-b" "us-central1-f"]
```

```
NAME                                       CAPACITY   ACCESSMODES   STATUS    CLAIM                  REASON    AGE       LABELS
pvc-237b0114-350a-11e6-964c-42010a800002   20Gi       RWO           Bound     default/datadir-rd-0             1m        failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-b
pvc-2382b0c2-350a-11e6-964c-42010a800002   20Gi       RWO           Bound     default/datadir-rd-1             1m        failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-f
pvc-23862681-350a-11e6-964c-42010a800002   20Gi       RWO           Bound     default/datadir-rd-2             1m        failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-a
```

davidopp commented on pkg/volume/util.go:
```diff
@@ -187,3 +192,47 @@ func GenerateVolumeName(clusterName, pvName string, maxLength int) string {
+		hash = h.Sum32()
+	}
+
+	zoneSlice := zones.List()
```
Does the spreading property depend on zones.List() returning the zones in the same order every time it's called? If so, do we know this will actually be true?

justinsb replied:
It does depend on it, but List() is sorted: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/sets/string.go#L167

This could probably use a comment though.

The case where it fails is if the set of zone changes while volumes are being created for a PetSet.  I think the most likely case here is that we grow a PetSet from N=1 -> N=3 some time _after_ initial creation, and in the meantime the user has also expanded their k8s cluster.  However, I don't really know what we can practically do better here in 1.3 (suggestions welcome).  We really should look at the set of PVCs and pick a zone based on the other allocations (like we would with pod anti affinity), but those scheduler-like changes seem impractical for 1.3 (and we know we want to do them later, which is why we are favoring the name-based hack over an annotation that might constrain us in future).

I hypothesize though that what we have here will be good enough in practice.  I don't think AWS or GCE have any regions with 5 zones.  I suspect most HA clusters will be launched into 3 zones, and then the set of zones will not change.

davidopp replied:
Yeah, that's fine, I was just curious. Thanks for the thorough explanation!

bprashanth replied:
Yeah i think we can simply document around scaling zones while a pet is scaling

bprashanth commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -914,6 +916,46 @@ func (aws *AWSCloud) List(filter string) ([]string, error) {
 }
 
+// getAllZones retrieves  a list of all the zones in which nodes are running
+// It currently involves querying all instances
```
how bad is this in large clusters in terms of delay? are we doing it per pvc, I'm guessing its not bad enough in terms of qps to put us into backoff?

justinsb replied:
I don't think creating PDs is a terribly common operation.  We used to list all instances every 10 seconds (but we no longer do that) so I think we should be OK.  Originally I did try caching, but now that we no longer poll Instances List, that becomes much harder.

I propose that we run this as-is for now, and then if/when we identify it as a problem we can add time-based caching.

Ideally the cloudprovider should have access to a watched set of Nodes, in which case this would be a trivial operation. This actually comes up in a number of places (e.g. LoadBalancer).  Maybe we do that in 1.4?

bprashanth replied:
SG

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -2066,6 +2068,38 @@ func (gce *GCECloud) List(filter string) ([]string, error) {
 }
 
+// GetAllZones returns all the zones in which nodes are running
+func (gce *GCECloud) GetAllZones() (sets.String, error) {
```
I'm guessing it's easier to do this than make it a private method in the same module, like we did for aws ?

justinsb replied:
Correct.  The GCE PD creation code is split between here and the volumes package; AWS has a cleaner separation of code.  So the zone is determine in the cloudprovider in AWS, and in the volumes code in GCE.

Note that this method is public, but is not part of the public interface.  i.e. it is public on GCECloud, not cloudprovider.Interface

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -2066,6 +2068,38 @@ func (gce *GCECloud) List(filter string) ([]string, error) {
+	// TODO: Parallelize, although O(zones) so not too bad (N <= 3 typically)
+	for _, zone := range gce.managedZones {
+		// We only retrieve one page in each zone - we only care about existence
+		listCall := gce.service.Instances.List(gce.projectID, zone)
```
is there a reson we're listing all instances here and only running instances above ? is it simpler to not account for instance state, since one can't guess if all instances will go down right after we allocate the pv anyway

justinsb replied:
I debated this, and am happy to change.  I tried to sum it up in "No filter: we assume that a zone is either used or unused".  My view was that here we should probably just keep it simple, and if an instance was starting up we would still count it as a zone in use.  The above method is constrained to exclude them, but I thought it probably simpler to just keep this method simple for now (we can always filter out instances that are shutting down later...)

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -2218,13 +2252,27 @@ func (gce *GCECloud) DeleteDisk(diskToDelete string) error {
+	var disk *gceDisk
+	var err error
+	if zone == "" {
+		disk, err = gce.getDiskByNameUnknownZone(name)
```
how bad would it be to always get by unknown zone, then match zone and throw an error/log? it might save us in the case where something fishy happens and we end up with cross matched zones

justinsb replied:
Ah... so this is subtle.  This is actually a workaround for #27656 and #27657.  The problem is that in the volumes admission controller the cloudprovider is not configured with multizone=true.  This would prevent getDiskByNameUnknownZone from working at all unless the PD was in the master zone.  However, the workaround here is to allow creation in a non-master zone by looking to this zone parameter, which is derived by the admission controller when the zone annotation is already set.

Also, we actually want to be moving away getDiskByNameUnknownZone and towards volumes that always have an associated zone on GCE, because volume names on GCE are unique per zone, not per region.  i.e. the correct specifier for a GCE PD is probably `<zone>/<name>`, not `<name>`.  Thankfully with dynamic provisioning we can just use the annotation (which is painful for a human but fine for code), and static provisioning never really made sense on a cloud anyway...

I guess this needs a comment!

bprashanth commented on pkg/volume/gce_pd/attacher_test.go:
```diff
@@ -342,6 +342,6 @@ func (testcase *testcase) DeleteDisk(diskToDelete string) error {
 }
 
-func (testcase *testcase) GetAutoLabelsForPD(name string) (map[string]string, error) {
+func (testcase *testcase) GetAutoLabelsForPD(name string, zone string) (map[string]string, error) {
```
how hard is a "name: foo-0" check zone type unittests? 

Optionally, you can actually create a shell petset (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/petset/fakes.go#L73), create a name identity mapper (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/petset/identity_mappers.go#L52), set the identity (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/petset/identity_mappers.go#L64), then pass it through the provisioner and check it's zone. That way I won't be able to get away with breaking the identity mapper without fixing this :)

justinsb replied:
I guess we could have two tests: one that tests that Petsets creates PVCs with names like `<prefix>-<ordinal>`, and the other that checks that PVCs with names like `<prefix>-<ordinal>` are round-robin-ed across zones.

bprashanth commented on pkg/volume/plugins.go:
```diff
@@ -49,6 +49,8 @@ type VolumeOptions struct {
 	// PV.Name of the appropriate PersistentVolume. Used to generate cloud
 	// volume name.
 	PVName string
+	// PVC.Name of the PersistentVolumeClaim, if we are auto-provisioning.
```
is this only set when we're auto-provisioning?

justinsb replied:
All of VolumeOptions is only used for auto-provisioning.  I guess "if we are auto-provisioning" is technically superfluous; I thought it was clearer but I guess it begs the question of when VolumeOptions is used other than auto-provisioning.

bprashanth commented on pkg/volume/util.go:
```diff
@@ -187,3 +192,47 @@ func GenerateVolumeName(clusterName, pvName string, maxLength int) string {
 }
+
+// ChooseZone implements our heuristics for choosing a zone for volume creation based on the volume name
+func ChooseZoneForVolume(zones sets.String, pvcName string) string {
```
can you document that if anyone creates volumes like foo-index they'll get spreading across zones? (just comment is fine)

justinsb replied:
Will do!

bprashanth commented on pkg/volume/util.go:
```diff
@@ -187,3 +192,47 @@ func GenerateVolumeName(clusterName, pvName string, maxLength int) string {
+
+	if pvcName == "" {
+		// We should always be called with a name; this shouldn't happen
+		glog.Warningf("No Name defined during volume create; choosing random zone")
```
please lower case 'Name'

bprashanth commented on pkg/volume/util.go:
```diff
@@ -187,3 +192,47 @@ func GenerateVolumeName(clusterName, pvName string, maxLength int) string {
+		lastDash := strings.LastIndexByte(pvcName, '-')
+		if lastDash != -1 {
+			petIDString := pvcName[lastDash+1:]
+			petID, err := strconv.ParseUint(petIDString, 10, 32)
```
can you please add a comment above where we insert the pet name identityt mentioning that it's used to detect zone? we should consider isolating this into a common function called `getOrdinality` or something that just returns an index for any given name that other parts of the system can reuse, but we can do that later and just put in a comment for now.

justinsb replied:
Sure.  I'm using the PVC name though - it seemed superfluous to add another annotation, and also like it would box us in for the future (we'd have to support it forever, whereas now we just have a heuristic that means that PetSet volumes do a sensible thing in most cases today, and in future they can do the optimal thing in all cases)

bprashanth commented on pkg/volume/util.go:
```diff
@@ -187,3 +192,47 @@ func GenerateVolumeName(clusterName, pvName string, maxLength int) string {
+				// We still hash the volume name, but only the base
+				hashString = pvcName[:lastDash]
+				glog.V(2).Infof("Detected PetSet-style volume name %q; index=%d", pvcName, index)
+			}
```
log the error at warningf, if any?

justinsb replied:
We expect an error though if the PVC name is not a PetSet created name (e.g. "my-cool-pvc").

erictune commented:
@quinton-hoole @bprashanth @pmorie This is marked as a release blocker and the author has a lot of other things on his plate.  Can one of the reviewers please give an LGTM if all major issues/questions have been addressed, and leave minor nits for follow up PRs, please.

justinsb committed:
Add comments & misc review fixes

Lots of comments describing the heuristics, how it fits together and the
limitations.

In particular, we can't guarantee correct volume placement if the set of
zones is changing between allocating volumes.

chrislovecnm commented:
@bgrant0607 @smarterclayton @saad-ali help ... this is another that is going to cause me heartburn :(  Can we get some momentum on review?

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -2066,6 +2068,48 @@ func (gce *GCECloud) List(filter string) ([]string, error) {
+		// volumes and the health of zones.
+
+		// Just a minimal set of fields - we only care about existence
+		listCall = listCall.Fields("items(name)")
```
@zmerlynn fyi another list call on fields, but I think this ones ok since it's just checking for exit code 0, essentially

zmerlynn replied:
This looks fine. The function could be a bit more pessimistic about which zones it uses if the `nodeInstancePrefix` I added in #27741 wasn't actually advisory, but right now we let people join arbitrary garbage to a master, so this sort of check is about all that you can do.

chrislovecnm commented:
@bprashanth @zmerlynn I gotta get this into a branch that I have to cut for a demo ... look good?

bprashanth commented:
I'm fine with this, I only had a bunch of nits. I'd really like the unittests I mentioned so I'm not allowed to silently break this feature, we don't have an e2e either, but I think we can address that through a follow up (filing issue).
